jpype/dbapi2.py

```diff
@@ -0,0 +1,1444 @@
+VARBINARY = JDBCType('VARBINARY', -3, 'getBytes', 'setBytes')
+VARCHAR = JDBCType('VARCHAR', 12, 'getString', 'setString')
+
+# Aliases required by DBAPI2
+STRING = JDBCType(['STRING', 'CHAR', 'NCHAR', 'NVARCHAR', 'VARCHAR', 'OTHER'], None,
+                  'getString', 'setString')
+TEXT = JDBCType(['TEXT', 'CLOB', 'LONGVARCHAR', 'LONGNVARCHAR', 'NCLOB', 'SQLXML'], None,
+                'getString', 'setString')
+BINARY = JDBCType(['BINARY', 'BLOB', 'LONGVARBINARY', 'VARBINARY'], -2,
+                  'getBytes', 'setBytes')
+NUMBER = JDBCType(['NUMBER', 'BOOLEAN', 'BIGINT', 'BIT', 'INTEGER', 'SMALLINT', 'TINYINT'], None,
+                  'getObject', 'setObject')
+FLOAT = _JDBCTypePrimitive(['FLOAT', 'REAL', 'DOUBLE'], 6,
+                           'getDouble', 'setDouble')
+DECIMAL = JDBCType(['DECIMAL', 'NUMERIC'], 3,
+                   'getBigDecimal', 'setBigDecimal')
+DATETIME = TIMESTAMP
+
+# Special types
+ASCII_STREAM = JDBCType(None, None, 'getAsciiStream', 'setAsciiStream')
+BINARY_STREAM = JDBCType(None, None, 'getBinaryStream', 'setBinaryStream')
+CHARACTER_STREAM = JDBCType(None, None, 'getCharacterStream', 'setCharacterStream')
+NCHARACTER_STREAM = JDBCType(None, None, 'getNCharacterStream', 'setNCharacterStream')
+URL = JDBCType(None, None, 'getURL', 'setURL')
+
+# The converters are defined in a customizer
+
+
+def _asPython(x):
+    return x._py()
+
+
+# This maps the types reported by the columns to the type used for the getter
+# and converter
+_default_map = {ARRAY: OBJECT, OBJECT: OBJECT, NULL: OBJECT,
+                REF: OBJECT, ROWID: OBJECT, RESULTSET: OBJECT,
+                TIME_WITH_TIMEZONE: OBJECT, TIMESTAMP_WITH_TIMEZONE: OBJECT,
+                NVARCHAR: STRING, CHAR: STRING,
+                NCHAR: STRING, VARCHAR: STRING, BINARY: BINARY,
+                BLOB: BINARY, LONGVARBINARY: BINARY, VARBINARY: BINARY,
+                NUMBER: NUMBER, BOOLEAN: BOOLEAN, BIGINT: BIGINT,
+                BIT: BIT, INTEGER: INTEGER, SMALLINT: SMALLINT,
+                TINYINT: TINYINT, FLOAT: FLOAT, REAL: REAL,
+                DECIMAL: DECIMAL, NUMERIC: NUMERIC,
+                DATE: DATE, TIMESTAMP: TIMESTAMP, TIME: TIME,
+                CLOB: STRING, NCLOB: STRING, STRING: STRING,
+                TEXT: STRING, SQLXML: STRING, LONGVARCHAR: STRING,
+                LONGNVARCHAR: STRING, DOUBLE: DOUBLE, OTHER: OBJECT
+                }
+
+_default_setters = {}
+
+_default_converters = {}
+
+_default_adapters = {}
+
+
+# Setters take (connection, meta, col, type) -> JDBCTYPE
+def SETTERS_BY_META(cx, meta, col, ptype):
+    """ Option for setters to use the metadata of the parameters.
+
+    On some databases this option is useless as they do not track parameter
+    types.  This method can be cached for faster performance when lots of
+    parameters.  Usually types can only be determined accurately on inserts
+    into defined columns.
+    """
+    return _default_map[_registry[meta.getParameterType(i + 1)]]
```

Comment:
where is i defined? Seems like broken/unused code to me

Reply:
Fixed.   It got renamed to col.